src/backend/joanie/core/flows/order.py

```diff
@@ -25,6 +25,184 @@ def _set_order_state(self, value):
     def _get_order_state(self):
         return self.instance.state
 
+    def _can_be_assigned(self):
+        """
+        An order can be assigned if it has an organization.
+        """
+        return self.instance.organization is not None
+
+    @state.transition(
+        source=enums.ORDER_STATE_DRAFT,
+        target=enums.ORDER_STATE_ASSIGNED,
+        conditions=[_can_be_assigned],
+    )
+    def assign(self, billing_address=None):
+        """
+        Transition order to assigned state.
+        """
+        if self.instance.product.price > 0 and billing_address:
+            Address = apps.get_model("core", "Address")  # pylint: disable=invalid-name
+            address, _ = Address.objects.get_or_create(
+                **billing_address,
+                owner=self.instance.owner,
+                defaults={
+                    "is_reusable": False,
+                    "title": f"Billing address of order {self.instance.id}",
+                },
+            )
+
+            # Create the main invoice
+            Invoice = apps.get_model("payment", "Invoice")  # pylint: disable=invalid-name
+            Invoice.objects.get_or_create(
+                order=self.instance,
+                total=self.instance.total,
+                recipient_address=address,
+            )
+
+        # taken from order.submit
+        ProductTargetCourseRelation = apps.get_model(  # pylint: disable=invalid-name
+            "core", "ProductTargetCourseRelation"
+        )
+        OrderTargetCourseRelation = apps.get_model("core", "OrderTargetCourseRelation")  # pylint: disable=invalid-name
+        for relation in ProductTargetCourseRelation.objects.filter(
+            product=self.instance.product
+        ):
+            order_relation = OrderTargetCourseRelation.objects.create(
+                order=self.instance,
+                course=relation.course,
+                position=relation.position,
+                is_graded=relation.is_graded,
+            )
+            order_relation.course_runs.set(relation.course_runs.all())
```

Comment:
WDYT, to move this logic in a method within the model ?
```suggestion
        self.instance.freeze_target_courses()
```